A cancelled run reports a failed required check, and a cancellation is not a defect

A run superseded by the next push leaves a **failed required check** on
the head it was cancelled for, and the reason is not a defect in the
branch.

## What was observed

On [bitcoin-core-rpc#133](https://github.com/btclib-org/bitcoin-core-rpc/pull/133),
`test: every job passed` reported **failure** while every suite job that
finished had passed. The run's own job list says why:

```text
cancelled  Run the suite on 3.10, ubuntu-latest
cancelled  Build the distribution and install it
cancelled  Run the suite on 3.14t, ubuntu-latest
cancelled  Run the suite on pypy3.11, ubuntu-latest
failure    test: every job passed
```

and the aggregate's own log carries nothing but its verdict:

```text
##[error]a job this gate depends on did not succeed
```

Four jobs were **cancelled**, not failed — `test.yml`'s concurrency
group with `cancel-in-progress`, doing exactly what it is for. The
aggregate carries `if: ${{ always() && ... }}`, so it runs anyway, and
its condition counts `cancelled` among the results that fail it.

## Why it matters

`CLAUDE.md` already warns that a branch's run **can be `cancelled`
rather than green**, and that `cancelled` is not `failure`. What it does
not say is that the aggregate turns the first into the second, and the
aggregate is a *required* check — so a cancellation becomes a merge
blocked by a red that describes no defect.

Worse, both runs report on the same head sha when the cancellation came
from a re-run or a branch update rather than from a new commit: the pull
request then carries a red and a green check of the same name, and the
red is what is read.

## Where it is, in this repository

`test.yml`'s `test-passed`, which reads the results into `RESULTS`
and exits 1 on `case " $RESULTS " in *" failure "* | *" cancelled "*`
(line 959). A different spelling of the same rule.

## What failing on `cancelled` is right about, and what it is not

Failing when a dependency was cancelled is defensible on its own terms:
the gate verified nothing, and reporting "every job passed" would be
false. The case that is wrong is the **concurrency** one, where the
cancellation means "a newer run is authoritative" rather than "this
tree is unproven".

`!cancelled()` in place of `always()` is the candidate: the aggregate
would be skipped when the run itself was cancelled, and a **skipped**
job satisfies a required check where a job that never runs blocks it —
which `REPOSITORY.md` already records for `integration.yml`'s `paths`
filter. Since a cancellation only happens because a superseding run
exists, the check that matters is the one that run produces.

That is a candidate and not a diagnosis: it wants testing on a
disposable branch before it is adopted, in the way the ruleset bypass
modes were. The failure mode to look for is the opposite one — an
aggregate that skips itself into satisfying a rule it should have held.

Observed in bitcoin-core-rpc and filed here because the same shape is
here too: the four repositories carry the same aggregate in three
spellings, so whatever is decided is decided once for all of them.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -922,17 +922,23 @@ jobs:
   # check, and macos.yml and windows.yml are two more workflows over the
   # same matrix shape.
   #
-  # `if: always()` because a job with unmet `needs` is skipped, and a
+  # `if: !cancelled()` because a job with unmet `needs` is skipped, and a
   # skipped check is not a failed one: without it a failing matrix cell
   # would leave this context reported as skipped, which satisfies a
-  # required check rather than blocking on it. Cancelled is a failure here
-  # too, a superseded run being no evidence either way
+  # required check rather than blocking on it. A cancellation of this
+  # run itself, though, is no evidence either way -- the concurrency
+  # group superseding it with a newer push -- and `always()` used to
+  # carry that into an explicit failure below despite the run's own
+  # cancellation already speaking for it. `!cancelled()` skips this job
+  # in that case instead, which satisfies the check the way any other
+  # skip does, while a job cancelled on its own rather than by the run's
+  # own cancellation still fails below
   test-passed:
     name: "test: every job passed"
     # and the draft condition every job above carries, so a draft
     # skips uniformly rather than tripping the skip check below -- the
     # closed one for the same reason, added at the trigger above
-    if: ${{ always() && !github.event.pull_request.draft && github.event.action != 'closed' }}
+    if: ${{ !cancelled() && !github.event.pull_request.draft && github.event.action != 'closed' }}
     needs:
       - changes
       - coverage

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2574,6 +2574,20 @@ release-notes length in the first place, and are still in
   both greps. `test: every job passed` gets an entry in the same pass,
   that being the neighbouring gap and the check a contributor sees red:
   it reproduces as nothing, reading the conclusions of the jobs above it.
+- **`test-passed` skips a cancelled run instead of failing it** (#274).
+  It carried `always()`, so a run the next push's concurrency group
+  already superseded — every needed job reporting `cancelled` — still
+  reached the `case " $RESULTS " in *" failure "* | *" cancelled "*)`
+  step and turned that into a required check with no defect behind it,
+  red beside the newer run's green on the same head sha. The job's own
+  comment had called this deliberate, "Cancelled is a failure here too,
+  a superseded run being no evidence either way" — the framing #273
+  disputes: the case that is wrong is the concurrency one, where the
+  cancellation means a newer run is authoritative rather than that this
+  tree is unproven. `always()` becomes `!cancelled()`; a skipped
+  required check satisfies branch protection the same as a passing one,
+  and the step below is unchanged, so a job cancelled on its own rather
+  than by the run's own cancellation still fails the gate as before.
 
 ## v0.8.0.2
 

